tests/rsa_test.c

```diff
@@ -0,0 +1,670 @@
+
+static int s_rsa_signing_roundtrip_pss_sha256_from_user(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    aws_cal_library_init(allocator);
+
+#if defined(AWS_OS_MACOS)
+    if (__builtin_available(macOS 10.12, *)) {
+        ASSERT_SUCCESS(s_rsa_signing_roundtrip_from_user(allocator, AWS_CAL_RSA_SIGNATURE_PSS_SHA256, NULL));
+    } else {
+        ASSERT_ERROR(
+            AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM,
+            s_rsa_signing_roundtrip_from_user(allocator, AWS_CAL_RSA_SIGNATURE_PSS_SHA256, NULL));
+    }
+#else
+    ASSERT_SUCCESS(s_rsa_signing_roundtrip_from_user(allocator, AWS_CAL_RSA_SIGNATURE_PSS_SHA256, NULL));
+#endif
+
+    aws_cal_library_clean_up();
+
+    return AWS_OP_SUCCESS;
+}
+AWS_TEST_CASE(rsa_signing_roundtrip_pss_sha256_from_user, s_rsa_signing_roundtrip_pss_sha256_from_user);
+
+static int s_rsa_getters(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    aws_cal_library_init(allocator);
+
+    struct aws_byte_buf private_key_buf;
+    ASSERT_SUCCESS(s_byte_buf_decoded_from_base64_cur(
+        allocator, aws_byte_cursor_from_c_str(TEST_PKCS1_RSA_PRIVATE_KEY_1024), &private_key_buf));
+    struct aws_rsa_key_pair *key_pair_private =
+        aws_rsa_key_pair_new_from_private_key_pkcs1(allocator, aws_byte_cursor_from_buf(&private_key_buf));
+
+    ASSERT_NOT_NULL(key_pair_private);
+    ASSERT_INT_EQUALS(128, aws_rsa_key_pair_block_length(key_pair_private));
+    ASSERT_INT_EQUALS(128, aws_rsa_key_pair_signature_length(key_pair_private));
+
+    struct aws_byte_buf priv_key;
+    ASSERT_SUCCESS(aws_rsa_key_pair_get_private_key(key_pair_private, AWS_CAL_RSA_KEY_EXPORT_PKCS1, &priv_key));
+    ASSERT_TRUE(priv_key.len > 0);
+
+    struct aws_byte_buf public_key_buf;
+    ASSERT_SUCCESS(s_byte_buf_decoded_from_base64_cur(
+        allocator, aws_byte_cursor_from_c_str(TEST_PKCS1_RSA_PUBLIC_KEY_1024), &public_key_buf));
+    struct aws_rsa_key_pair *key_pair_public =
+        aws_rsa_key_pair_new_from_public_key_pkcs1(allocator, aws_byte_cursor_from_buf(&public_key_buf));
+
+    ASSERT_INT_EQUALS(128, aws_rsa_key_pair_block_length(key_pair_public));
+    ASSERT_INT_EQUALS(128, aws_rsa_key_pair_signature_length(key_pair_public));
+
+    struct aws_byte_buf pub_key;
+    ASSERT_SUCCESS(aws_rsa_key_pair_get_public_key(key_pair_public, AWS_CAL_RSA_KEY_EXPORT_PKCS1, &pub_key));
+    ASSERT_TRUE(pub_key.len > 0);
+
+    aws_rsa_key_pair_release(key_pair_private);
+    aws_rsa_key_pair_release(key_pair_public);
+    aws_byte_buf_clean_up_secure(&private_key_buf);
+    aws_byte_buf_clean_up_secure(&public_key_buf);
+    aws_byte_buf_clean_up_secure(&priv_key);
+    aws_byte_buf_clean_up_secure(&pub_key);
+
+    aws_cal_library_clean_up();
+
+    return AWS_OP_SUCCESS;
+}
+AWS_TEST_CASE(rsa_getters, s_rsa_getters);
+
+static int s_rsa_private_pkcs1_der_parsing(struct aws_allocator *allocator, void *ctx) {
```

Comment:
Bad encoded test cases to consider:
1. [Negative modulus](https://github.com/aws/aws-lc/blob/main/crypto/rsa_extra/rsa_test.cc#L329)
2. [1 exponent](https://github.com/aws/aws-lc/blob/main/crypto/rsa_extra/rsa_test.cc#L357)

Reply:
do you know how prevalent negative modulus is in the wild? our der encoder/decoder has some limitations around negative number handling and will currently throw errors if negative numbers are encountered.

Reply:
I don't have any data for how prevalent they are, at least one [chrome user hit this issue](https://crbug.com/532048). [AWS-LC doesn't allow it](https://github.com/aws/aws-lc/blob/main/crypto/rsa_extra/rsa_test.cc#L741-L744) and no one has complained to us (yet).